Add static method

LoiNguyenCS opened the pull request.

LoiNguyenCS commented:
Professor,

When I was writing the paper for Specimin, I realized that I forgot to handle static method calls in the input program, which is quite embarrassing. This pull request addresses that by adding the necessary code for unsolved static method calls. Additionally, I noticed the existence of the `tempDirJDK` folder, which I had created a while ago for bug testing purposes. I believe it was accidentally added at some point in the past, so this PR includes the removal of that folder.

In this updated version of Specimin, wildcard imports are no longer allowed. The reason is that we need to distinguish between class and method names in unsolved method calls, such as `A.c()`. The only reliable way to determine whether `A` refers to a class or a method is to look at the import statements. 

I think the logic of this solution might not apply if there is a method or class instance with the same name as a class among the import statements. But I think the odd is low, since usually people don't name methods or class instances with a capitalized name like they do for classes.

kelloggm commented on src/test/resources/unsolvedstaticmethod/expected/com/example/Class.java:
```diff
@@ -0,0 +1,8 @@
+
+public class Class {
+
+    public ProcessReturnType process() {
```
Shouldn't this be `static`?

kelloggm reviewed (changes requested):
Nice catch! Please also add a test for a more complex static call (e.g., one where the programmer uses the fully-qualified name of the class).

LoiNguyenCS commented:
Professor,

I've made some changes to the code.

My initial implementation was incorrect. The problem was somewhat concealed because I named my test class `Class`, coincidentally matching some built-in Java classes. Thus, the program ran without triggering any error.

I've also introduced a more complex method call in the test file.

Please take a look. Thank you.

LoiNguyenCS commented on src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:
```diff
@@ -1216,7 +1219,7 @@ public static boolean isAClassPath(String potentialClassPath) {
    * @return the corresponding instance of UnsolvedClass
    */
-  public static UnsolvedClass getSimpleSyntheticClassFromFullyQualifiedName(
+  public UnsolvedClass getSimpleSyntheticClassFromFullyQualifiedName(
```
Ah, I was trying another approach earlier. I will re-add `static`

LoiNguyenCS commented on src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:
```diff
@@ -1254,12 +1294,17 @@ public static boolean unsolvedAndNotSimple(MethodCallExpr method) {
+    // As this code involves complex string operations, we'll use a method call as an example,
+    // following its progression through the code.
+    // Suppose this is our method call: com.example.MyClass.process()
+    // At this point, our method call become: com.example.MyClass.process
```
Thank you for pointing this out. I've changed the replace call into a substring call.

LoiNguyenCS commented:
Professor,

I have updated the codes and the test file.

LoiNguyenCS commented:
I don't know why the CI is not running. I think this is a [bug](https://github.com/orgs/community/discussions/31587) of GitHub. But the solutions discussed in that bug report don't seem to be related to our case.

kelloggm commented on src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:
```diff
@@ -1254,43 +1294,58 @@ public static boolean unsolvedAndNotSimple(MethodCallExpr method) {
    *
-   * @param method the method call to be taken as input
+   * @param methodCall the method call to be taken as input
+   * @param methodAgruments the list of agruments for this method call
```
typo: "arguments"

kelloggm commented on src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:
```diff
@@ -1254,43 +1294,58 @@ public static boolean unsolvedAndNotSimple(MethodCallExpr method) {
+   * corresponding synthetic class.
    *
-   * @param method the method call to be taken as input
+   * @param methodCall the method call to be taken as input
```
This string needs to be a particular format. I think it needs to be a static invocation: in particular, it must not have a receiver, since you use a substring call to find the first `(` and expect the result to be the beginning of the arguments list. If there was a receiver like the following, this would fail:
"foo.getBar().methodCall(...)" - if we passed this whole string, the substring call below will only find "foo.getBar", not "foo.getBar().methodCall", as we intend.

So, please give the exact specification that you expect here in this Javadoc.

kelloggm reviewed (changes requested):
I'm still concerned about the string manipulation.

Also, I'm not sure that this code will handle [static imports](https://docs.oracle.com/javase/8/docs/technotes/guides/language/static-import.html) correctly. Can you also add a test that statically imports a method and make sure that the code handles it properly?

LoiNguyenCS commented:
Professor,

> I'm still concerned about the string manipulation.

I have renamed the related methods and added clarification in the Javadoc to make sure that the expected input will be passed to this string manipulation call.

> Also, I'm not sure that this code will handle [static imports](https://docs.oracle.com/javase/8/docs/technotes/guides/language/static-import.html) correctly. Can you also add a test that statically imports a method and make sure that the code handles it properly?

Unfortunately, I wasn't aware of this Java feature until today. Updating Specimin for static imports might involve extensive modifications, such as changing how it reads import statements and updates `classAndPackageMap`. I suggest addressing this in a separate PR, ideally after resolving issue #41.
`

kelloggm commented:
> Updating Specimin for static imports might involve extensive modifications, such as changing how it reads import statements and updates classAndPackageMap. I suggest addressing this in a separate PR, ideally after resolving issue https://github.com/kelloggm/specimin/issues/41.

They are rare, so that's fine. Please open an issue for this so we don't forget.

kelloggm commented on src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:
```diff
@@ -1293,14 +1293,15 @@ public boolean unsolvedAndNotSimple(MethodCallExpr method) {
    *
-   * @param methodCall the method call to be taken as input
-   * @param methodAgruments the list of agruments for this method call
+   * @param methodCall the method call to be used as input
```
I would still prefer that the `@param` Javadoc here explains exactly the requirements on `methodCall` in string terms.

kelloggm replied:
That is, "must contain one or more dot separated identifiers, followed by a single pair of parentheses containing arguments" or something similar

kelloggm reviewed (approved):
When the requested changes are implemented, you can merge without requiring another review from me.
